server/src/mds3.gdc.js

```diff
@@ -1282,94 +1285,190 @@ returns:
 */
 
 export async function querySamples_gdcapi(q, twLst, ds, geneTwLst) {
+	// step 1: separate survival vs non-survival terms
+	// survival terms requires different querying method. separate it from other dictionary terms
+
+	const survivalTwLst = [], // survival terms
+		dictTwLst = [] // non-survival dict terms
+	for (const tw of twLst) {
+		if (tw.term.type == 'survival') {
+			survivalTwLst.push(tw)
+		} else {
+			const t = ds.cohort.termdb.q.termjsonByOneid(tw.term.id)
+			if (t) dictTwLst.push({ term: t, q: tw.q })
+		}
+	}
+
+	let byTermId = {},
+		samples = [] // returned data structures. both survival and other dict term data will be combined for return
+
+	// step 2: querying dict term data
+
 	if (q.isHierCluster) {
-		/*
-		running gene exp clustering
-		must only restrict to cases with exp data
-		but not by mutated cases anymore, thus geneTwLst should not be used (and not supplied)
+		// running gene exp clustering, must only restrict to cases with exp data, but not by mutated cases anymore, thus geneTwLst should not be used (and not supplied)
+		;[byTermId, samples] = await querySamplesTwlstForGeneexpclustering(q, dictTwLst, ds)
+	} else {
+		// not in gene exp clustering mode
+		;[byTermId, samples] = await querySamplesTwlstNotForGeneexpclustering(q, dictTwLst, ds, geneTwLst)
+	}
+
+	// step 3: querying survival data
+
+	if (survivalTwLst.length) {
+		await querySamplesSurvival(q, survivalTwLst, ds, samples, geneTwLst)
+	}
+
+	return { byTermId, samples }
+}
+
+async function querySamplesSurvival(q, survivalTwLst, ds, samples, geneTwLst) {
+	// build survival api filter
+	const filter = { op: 'and', content: [] }
+	if (q.filter0) {
+		filter.content.push(q.filter0)
+	}
+
+	if (geneTwLst) q.isoforms = mapGenes2isoforms(geneTwLst, q.genome)
+
+	addSsmIsoformRegion4filter(filter.content, q, 'survival')
+
+	delete q.isoforms
+
+	if (q.set_id) {
+		if (typeof q.set_id != 'string') throw '.set_id value not string'
+		filter.content.push({
+			op: 'in',
+			content: {
+				field: 'cases.case_id',
+				value: [q.set_id]
+			}
+		})
+	}
+	if (q.filterObj) {
+		const g = filter2GDCfilter(q.filterObj)
+		if (g) filter.content.push(g)
+	}
+	addTid2value_to_filter(q.tid2value, filter.content, ds)
+
+	// the survival term itself is not used in api query, since there's just one type of survival data from gdc and no need to distinguish
+	const { host, headers } = ds.getHostHeaders(q)
+	const re = await ky
+		// hardcoded url in production environment. TODO check if works at all for qa environments
+		.post('https://portal.gdc.cancer.gov/auth/api/v0/analysis/survival', {
```

Comment:
@siosonel please assess this

Reply:
@xzhou82  The same URL should work in qa environments, it'll just point to different IP addresses depending on the VPN user

Reply:
thanks replaced with `host.rest` and works